src/concurrent_cache.h

```diff
@@ -70,6 +95,10 @@ class ConcurrentCache
     if ( x.miss() ) {
       try {
         valuePromise.set_value(f());
+        {
+          std::unique_lock<std::mutex> l(lock_);
+          impl_.increaseSize(CostEstimation::cost(x.value().get()));
```

Comment:
I think that only the call to `lru_cache::increaseSize()` should be protected by a lock. Even though cost computation should be fast it's not up to `ConcurrentCache` to assume that it really is.